src/librustc_mir/borrow_check/diagnostics/region_errors.rs

```diff
@@ -53,13 +54,69 @@ impl ConstraintDescription for ConstraintCategory {
     }
 }
 
-#[derive(Copy, Clone, PartialEq, Eq)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
 enum Trace {
     StartRegion,
     FromOutlivesConstraint(OutlivesConstraint),
     NotVisited,
 }
 
+/// A collection of errors encountered during region inference. This is needed to efficiently
+/// report errors after borrow checking.
+crate type RegionErrors<'tcx> = smallvec::SmallVec<[RegionErrorKind<'tcx>; 4]>;
+
+#[derive(Clone, Debug)]
+crate enum RegionErrorKind<'tcx> {
+    /// An error for a type test: `T: 'a` does not live long enough
+    TypeTestDoesNotLiveLongEnough {
+        /// The span of the type test
+        span: Span,
+        /// The generic type of the type test
+        generic: GenericKind<'tcx>,
+    },
+
+    /// A generic bound failure for a type test
+    TypeTestGenericBoundError {
+        /// The span of the type test
+        span: Span,
+        /// The generic type of the type test
+        generic: GenericKind<'tcx>,
+        /// The lower bound region
+        lower_bound_region: ty::Region<'tcx>,
+    },
+
+    /// An unexpected hidden region for an opaque type
+    UnexpectedHiddenRegion {
+        /// The def id of the opaque type
+        opaque_type_def_id: DefId,
+        /// The hidden type
+        hidden_ty: Ty<'tcx>,
+        /// The unexpected region
+        member_region: ty::Region<'tcx>,
+    },
+
+    /// Any other lifetime error
+    RegionError {
+        /// The origin of the region
+        fr_origin: NLLRegionVariableOrigin,
+        /// The region that should outlive `shorter_fr`
+        longer_fr: RegionVid,
+        /// The region that should be shorter, but we can't prove it
+        shorter_fr: RegionVid,
+    },
+
+    /// The same as `RegionError`, but this is an unreported error. We currently only report the
```

Comment:
I think this should be merged with `RegionError` and have an `error_reported: bool` flag. The use sites of this error would make this easy to use this way and the documentation you currently have on this variant could live on the bool flag.